File: packages/template/scripts/vcoo-whatsapp-pair.py
#!/usr/bin/env python3
"""VCOO WhatsApp pairing — runs bridge.js --pair-only or vcoo-whatsapp-pair.js."""
import json, os, subprocess, sys, time

HERMES_HOME = os.path.expanduser("~/.hermes")
BRIDGE_DIR = os.path.join(HERMES_HOME, "hermes-agent", "scripts", "whatsapp-bridge")

phone = sys.argv[1] if len(sys.argv) > 1 else None

NODE_SCRIPT = os.path.join(HERMES_HOME, "scripts", "vcoo", "vcoo-whatsapp-pair.js")
BRIDGE_JS = os.path.join(BRIDGE_DIR, "bridge.js")

if os.path.isfile(NODE_SCRIPT):
    cmd = ["node", NODE_SCRIPT]
    if phone:
        cmd += ["--phone", phone]
elif os.path.isfile(BRIDGE_JS):
    BRIDGE_NM = os.path.join(BRIDGE_DIR, "node_modules")
    if not os.path.isdir(BRIDGE_NM):
        print(json.dumps({"status": "installing_whatsapp"}))
        sys.stdout.flush()
        subprocess.run(["npm", "install", "--no-audit", "--no-fund", "--loglevel=error"],
                       cwd=BRIDGE_DIR, capture_output=True, timeout=180)
    env = os.environ.copy()
    env.update({"WHATSAPP_MODE": "bot", "WHATSAPP_ALLOWED_USERS": "*"})
    if phone:
        env["WHATSAPP_ALLOWED_USERS"] = phone
    cmd = ["node", BRIDGE_JS, "--pair-only", "--pair-json"]
    env_cmd = cmd
    try:
        proc = subprocess.Popen(env_cmd, cwd=BRIDGE_DIR, stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE, text=True, bufsize=1, env=env)
    except FileNotFoundError:
        print(json.dumps({"error": "node not found"}))
        sys.exit(1)
    for line in proc.stdout:
        line = line.strip()
        if not line:
            continue
        try:
            event = json.loads(line)
        except json.JSONDecodeError:
            continue
        ev = event.get("event", "")
        if ev == "qr":
            print(json.dumps({"qr": event.get("qr", ""), "status": "waiting"}))
            sys.stdout.flush()
        elif ev == "connected":
            print(json.dumps({"status": "connected", "user": event.get("user", {})}))
            sys.stdout.flush()
            proc.terminate()
            sys.exit(0)
        elif ev == "error":
            print(json.dumps({"status": "error", "error": event.get("error", "")}))
            sys.stdout.flush()
            proc.terminate()
            sys.exit(1)
    proc.wait()
    sys.exit(0)
else:
    print(json.dumps({"error": "WhatsApp bridge not found. Install Hermes first."}))
    sys.exit(1)

try:
    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE, text=True, bufsize=1)
except FileNotFoundError:
    print(json.dumps({"error": "node not found"}))
    sys.exit(1)

waiting_for = "pairing_code" if phone else "qr"
found = False
start_time = time.time()

for line in proc.stdout:
    line = line.strip()
    if not line:
        continue
    try:
        event = json.loads(line)
    except json.JSONDecodeError:
        continue

    ev = event.get("event", "")

    if ev == "qr" and not phone:
        output = {"qr": event.get("qr", ""), "status": "waiting"}
        print(json.dumps(output))
        sys.stdout.flush()
        found = True

    elif ev == "pairing_code":
        output = {"pairing_code": event.get("code", ""), "phone": event.get("phone", ""), "status": "waiting"}
        print(json.dumps(output))
        sys.stdout.flush()
        found = True

    elif ev == "connected":
        output = {"status": "connected", "user": event.get("user", {})}
        print(json.dumps(output))
        sys.stdout.flush()
        proc.terminate()
        sys.exit(0)

    elif ev == "installing":
        print(json.dumps({"status": "installing_whatsapp"}))
        sys.stdout.flush()

    elif ev == "error":
        print(json.dumps({"status": "error", "error": event.get("error", ""), "message": event.get("message", "")}))
        sys.stdout.flush()
        proc.terminate()
        sys.exit(1)

    if found and time.time() - start_time > 120:
        break

proc.wait(timeout=60)

File: packages/template/scripts/vcoo-whatsapp-pair.js
#!/usr/bin/env node
/* VCOO WhatsApp pairing — supports QR code and pairing code modes.
 * Usage:
 *   node vcoo-whatsapp-pair.js                          # QR mode (default)
 *   node vcoo-whatsapp-pair.js --phone +1234567890      # Pairing code mode
 * Outputs JSON lines: {event, qr?, code?, error?}
 */
const path = require('path');
const fs = require('fs');

// Determine bridge directory (relative to Hermes home)
const HERMES_HOME = process.env.HERMES_HOME || path.join(require('os').homedir(), '.hermes');
const BRIDGE_DIR = path.join(HERMES_HOME, 'hermes-agent', 'scripts', 'whatsapp-bridge');
const SESSION_DIR = process.env.SESSION_DIR || path.join(HERMES_HOME, 'whatsapp-session', 'main');
const BAIL_EYS = path.join(BRIDGE_DIR, 'node_modules', '@whiskeysockets', 'baileys');

function emit(ev) {
  process.stdout.write(JSON.stringify({ ts: Date.now(), ...ev }) + '\n');
}

// Parse args
const args = process.argv.slice(2);
const phoneIdx = args.indexOf('--phone');
const phoneNumber = phoneIdx >= 0 ? args[phoneIdx + 1] : null;

async function main() {
  // Ensure Baileys is installed
  if (!fs.existsSync(path.join(BAIL_EYS, 'package.json'))) {
    emit({ event: 'installing' });
    const { execSync } = require('child_process');
    execSync('npm install --no-audit --no-fund --loglevel=error', {
      cwd: BRIDGE_DIR,
      stdio: 'pipe',
      timeout: 120000,
    });
  }

  const Baileys = require('@whiskeysockets/baileys');
  const { makeWASocket, useMultiFileAuthState, DisconnectReason } = Baileys;
  const P = require('pino');

  fs.mkdirSync(SESSION_DIR, { recursive: true });
  const { state, saveCreds } = await useMultiFileAuthState(SESSION_DIR);

  const sock = makeWASocket({
    auth: state,
    printQRInTerminal: false,
    logger: P({ level: 'silent' }),
    browser: ['VCOO', 'Chrome', ''],
    syncFullHistory: false,
    markOnlineOnConnect: false,
  });

  let paired = false;

  sock.ev.on('creds.update', saveCreds);

  sock.ev.on('connection.update', async (update) => {
    const { connection, lastDisconnect, qr } = update;

    if (qr && !paired && !phoneNumber) {
      emit({ event: 'qr', qr });
    }

    if (connection === 'open') {
      paired = true;
      const user = sock.authState.creds.me;
      emit({ event: 'connected', user });
      setTimeout(() => process.exit(0), 500);
    }

    if (connection === 'close') {
      const statusCode = lastDisconnect?.error?.output?.statusCode;
      if (statusCode === DisconnectReason.loggedOut) {
        emit({ event: 'error', error: 'logged_out' });
      } else if (!paired) {
        emit({ event: 'error', error: 'connection_closed', code: statusCode });
      }
      process.exit(paired ? 0 : 1);
    }
  });

  // If phone number provided, request pairing code
  if (phoneNumber) {
    // Wait a moment for socket to initialize
    await new Promise(r => setTimeout(r, 2000));
    try {
      const code = await sock.requestPairingCode(phoneNumber);
      emit({ event: 'pairing_code', code, phone: phoneNumber });
    } catch (err) {
      emit({ event: 'error', error: 'pairing_code_failed', message: err.message });
    }
  }
}

main().catch((err) => {
  emit({ event: 'error', error: err.message });
  process.exit(1);
});
